actions/actions_whitebox_test.go

```diff
@@ -0,0 +1,176 @@
+		changes, err := fxt.WorkItems[0].ChangeSet(wiCopy)
+		require.Nil(t, err)
+		require.Len(t, changes, 2)
+		// we intentionally test the order here as the code under test needs
+		// to be expanded later, supporting more changes and this is an
+		// integrity test on the current impl.
+		require.Equal(t, "system.state", changes[0].AttributeName)
+		require.Equal(t, "new", changes[0].NewValue)
+		require.Equal(t, "open", changes[0].OldValue)
+		require.Equal(t, "system.boardcolumns", changes[1].AttributeName)
+		require.Equal(t, fxt.WorkItems[0].Fields["system.boardcolumns"], changes[1].NewValue)
+		require.Equal(t, wiCopy.Fields["system.boardcolumns"], changes[1].OldValue)
+	})
+}
+
+func (s *ActionSuite) TestActionExecution() {
+	fxt := tf.NewTestFixture(s.T(), s.DB, tf.CreateWorkItemEnvironment(), tf.WorkItems(2))
+	require.NotNil(s.T(), fxt)
+	require.Len(s.T(), fxt.WorkItems, 2)
+	userID := fxt.Identities[0].ID
+
+	s.T().Run("by Old New", func(t *testing.T) {
+		fxt.WorkItems[0].Fields["system.state"] = "new"
+		fxt.WorkItems[0].Fields["system.boardcolumns"] = []interface{}{"bcid0", "bcid1"}
+		newVersion := createWICopy(*fxt.WorkItems[0], "open", []interface{}{"bcid0", "bcid1"})
+		afterActionWI, changes, err := ExecuteActionsByOldNew(s.Ctx, s.GormDB, userID, fxt.WorkItems[0], newVersion, map[string]string{
+			"Nil": "{ noConfig: 'none' }",
+		})
+		require.Nil(t, err)
+		require.Len(t, changes, 0)
+		require.Equal(t, "open", afterActionWI.(workitem.WorkItem).Fields["system.state"])
+	})
+
+	s.T().Run("by ChangeSet", func(t *testing.T) {
+		fxt.WorkItems[0].Fields["system.state"] = "new"
+		fxt.WorkItems[0].Fields["system.boardcolumns"] = []interface{}{"bcid0", "bcid1"}
+		newVersion := createWICopy(*fxt.WorkItems[0], "open", []interface{}{"bcid0", "bcid1"})
+		contextChanges, err := fxt.WorkItems[0].ChangeSet(newVersion)
+		require.Nil(t, err)
+		afterActionWI, changes, err := ExecuteActionsByChangeset(s.Ctx, s.GormDB, userID, newVersion, contextChanges, map[string]string{
+			"Nil": "{ noConfig: 'none' }",
+		})
+		require.Nil(t, err)
+		require.Len(t, changes, 0)
+		require.Equal(t, "open", afterActionWI.(workitem.WorkItem).Fields["system.state"])
+	})
+
+	s.T().Run("unknown rule", func(t *testing.T) {
+		fxt.WorkItems[0].Fields["system.state"] = "new"
+		fxt.WorkItems[0].Fields["system.boardcolumns"] = []interface{}{"bcid0", "bcid1"}
+		newVersion := createWICopy(*fxt.WorkItems[0], "open", []interface{}{"bcid0", "bcid1"})
+		contextChanges, err := fxt.WorkItems[0].ChangeSet(newVersion)
+		require.Nil(t, err)
+		_, _, err = ExecuteActionsByChangeset(s.Ctx, s.GormDB, userID, newVersion, contextChanges, map[string]string{
+			"unknownRule": "{ noConfig: 'none' }",
+		})
+		require.NotNil(t, err)
+	})
+
+	s.T().Run("sideffects", func(t *testing.T) {
+		fxt.WorkItems[0].Fields["system.state"] = "new"
+		fxt.WorkItems[0].Fields["system.boardcolumns"] = []interface{}{"bcid0", "bcid1"}
```

Comment:
You shouldn't store strings in the boardcolumns field. These are UUIDs.

Reply:
No! This is the JSON-deserialized field value. It has strings. Otherwise the code would fail on production. But it is true that this is very inconsistent. In some areas, strings are used, and in others UUIDs are used. This is the same issue as with the dreaded interface type insecurity: we should avoid using interface{} at any cost. It introduces a lot of possibly inconsistency.

Reply:
Trust me, please. Use `[]interface{}{uuid.NewV4().String(), uuid.NewV4().String()}` if you want to store strings but please don't just store `bcid0` which is just crap and will later fail in all my tests.

Reply:
If this fails, then the code is broken. The JSON is supposed to hold any value.